### ISSUE TITLE
Project Roadmap to v1.0

### ISSUE THREAD

**Issue by rrrene** [rrrene/credo#15]
I've seen promising Elixir projects getting a good start and then they somehow fade away. Maintainers get dragged away, no clear path was set and then the projects started meandering, not updating when Elixir reached 1.0, no longer responding to issues, the project merely survives because "this idea is great" ...

I don't want that to happen here and so I am continuing [a tradition started in the ElixirStatus project](https://github.com/rrrene/elixirstatus-web/issues/21): **Below you find a rough "Roadmap to 1.0".**

Please be aware of two things:
- **all of this is up for discussion**, which is why I publish this as a GitHub Issue instead of a blog post
- the version numbers are not set in stone, they are just meant as rough "milestones"

With that in mind, here we go:

## ~~0.1.x - getting off the ground~~

- fixing bugs
- improving the README
- fixing the UI (esp. wording)

## ~~0.2.x - more checks and utility~~

- more checks
  I still have a list of checks I want to implement and think it is good to "get this done" in order to appeal to a wider audience
- error codes
  I have an idea for a simple yet effective way to include error codes (or exit statuses) so that you can differentiate between issues in your CI builds, to e.g. react differently to Warnings than to Readability issues. By default, ~~"Software design" issues like "TODO: ..." comments will not result in an error, but you will be able to customize this to your needs.~~ there is an error when you would see issues in the output. But you can customize that on a per check basis!

## ~~0.3.x - formats and ignore options~~

- there is a "formatter API" to support implementation of different formatters (file list, Flycheck, ...)
- we have a way to disable Credo for specific functions

## ~~0.4.x - community~~

- the idea is that at this point, all checks I personally want to see in Credo have been implemented, so this will be a great moment to "let go" a bit and encourage community contributions even more
- to achieve this I want to make it dead simple to (a) create your own closed-source checks, but at the same time make it super-easy to contribute your checks back to the community
- all the utility stuff living under the `Credo.Code.*` namespace will be polished to provide a kind of toolbox for creating checks

## ~~0.9.x~~

- introduce `Credo.Execution` struct

## 0.10.x - almost final changes - :zap: (we are here)

- refactor internals to support post 1.0 changes
- fix known bugs

## 0.11.x - final changes

- finalize structure of `Credo.Execution` struct
- freeze the public API
- improve docs of the public API
- according to this vaporware roadmap, this will be the last release before 1.0

---

Lastly, there are things this project could benefit from which are on my radar but did not fit well  into the above list.

This includes trivial things like a nice project homepage (because a project about educating developers should at least be able to educate them about its existence, right?), but this also includes specificly non-trivial things like speeding Credo up on medium to large repos (e.g. through smarter code, caching, you name it) or adding a way to ignore certain issues for certain lines of code (as suggested in #5).

---

**TL;DR** There is a plan. Please feel free to provide encouraging words and critical thoughts below.

**Comment by Ramblurr** [comment]
Are there any plans to implement 'fixing" of format errors/warning? I'd normally call that a code formatter, but we're already using that term here.

**Comment by rrrene** [comment]
I guess you mean something like auto-correct? 

I am not completely against it, but I also think that one might become pandora's box if we tackle it too soon. Pretty sure this would be a post-1.0 item ...

Any thoughts on this?

**Comment by Tuxified** [comment]
At "0.3.0" a Formatter API is mentioned. I'm asking because I'd like to add Credo to CodeClimate engines, for which it needs JSON output.

**Comment by rrrene** [comment]
We do not have JSON output yet. There is also no real idea of a Formatter API I have to admit. I will have to update this issue. Thx for reporting! :+1:

P.S.: There is, I think, already a CodeClimate Engine which supports Dogma and Credo. Read something about it on the Twitter, but I have not followed that project.

**Comment by devonestes** [comment]
Hey - happy to see a roadmap to 1.0! Thanks for taking the time to put it together.

I was just looking at the current issues and I was wondering what would be most helpful for me to jump in on next. I have a feeling the formatter API is something you'd like to tackle, but if there are any priority issues or things that you know are blockers for getting to 1.0, then just let me know and I'll chip in where I can.

**Comment by rrrene** [comment]
I am a bit torn whether or not I want JSON support "at all costs" for `v1.0`. It would sure be nice, but to be honest, I do not have an elegant idea how to do it, yet.

With regards to the magic `v1.0` milestone, bug hunting is an equally important endeavour. :+1:
